modules/starfish/2.0/starfish2.smk

```diff
@@ -0,0 +1,247 @@
+        name_dict = {names[i]: callers[i] for i in range(len(callers))}
+        outdir = os.path.dirname(input.complete)
+        for src in glob.glob(outdir + "/[A-Z]*.vcf.gz*"): 
+            bname = os.path.basename(src)
+            lhs, rhs = bname.split(".", 1)
+            lhs = "_".join(lhs)
+            for key in list(name_dict.keys()): 
+                lhs = lhs.replace(key, name_dict[key])
+            dest = os.path.join(outdir, ".".join([lhs, rhs]))
+            os.rename(src, dest) 
+
+
+def _starfish_get_output(wildcards): 
+    # Get the checkpoint output directory
+    CFG = config["lcr-modules"]["starfish"]
+    checkpoint_output = os.path.dirname(checkpoints._starfish_rename_output.get(**wildcards).output.renamed)
+    # Rename each VCF with the variant caller names
+    vcfs = expand(
+        CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{vcf_name}.vcf.gz",
+        seq_type = wildcards.seq_type,
+        genome_build = wildcards.genome_build, 
+        tumour_id = wildcards.tumour_id, 
+        normal_id = wildcards.normal_id, 
+        pair_status = wildcards.pair_status, 
+        vcf_name = glob_wildcards(os.path.join(checkpoint_output, "{vcf_name}.vcf.gz")).vcf_name
+        )
+    return vcfs
+
+# Create a union VCF
+rule _starfish_union: 
+    input: 
+        _starfish_get_output
+    output: 
+        union = CFG["dirs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz", 
+        union_tbi = CFG["dirs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz.tbi"
+    log:
+        stdout = CFG["logs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.stdout.log",
+        stderr = CFG["logs"]["starfish"] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.stderr.log"
+    params: 
+        options = CFG["options"]["starfish_union"]
+    conda: 
+        CFG["conda_envs"]["bcftools"]
+    threads:
+        CFG["threads"]["starfish_union"]
+    resources:
+        **CFG["resources"]["starfish_union"]
+    shell: 
+        op.as_one_line("""
+        bcftools merge {params.options} --force-samples 
+        -Oz -o {output.union} 
+        {input} && 
+        tabix -p vcf {output.union}
+        """)
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _starfish_output_vcf:
+    input:
+        vcf = CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{vcf_name}.vcf.gz"
+    output:
+        vcf = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{vcf_name}.vcf.gz", 
+        tbi = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{vcf_name}.vcf.gz.tbi"
+    wildcard_constraints: 
+        vcf_name = "\S+(?<!union)"
```

Comment:
What does this actually do?

Reply:
This is the replacement for the "not_union_vcf" variable. It's a negative lookbehind assertion, meaning this value is restricted to those that don't end with "union"

Reply:
I figured. Seems overly complex. Why can't the same rule output the union and non-union VCFs? I'm sure there's a good reason but I can't glean it from this, mostly because I don't understand much of the other stuff going on within checkpoints.